### PR TITLE
refactor(pg): store nil map as empty {}

### PR DESCRIPTION
**Pull request by janisz**
## Description

Currently we store nil map as `'null'` (string). In next version of pgx the semantic has changed and it will be stored as `null`. 
This PR seal this behaviour and whenever we would like to store nil map it will be replaced with `{}` (empty but not nil map).

`TestIndex/TestMapHighlights/key_does_not_exist` shows the difference between having `'null'` and `null` in JSON field and null always returns false so negative queries will not work correctly.  

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

CI

### CODE REVIEW

**Comment by janisz** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6929** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6929" target="_blank">Graphite</a>  👈
    * **PR #6915** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6915" target="_blank">Graphite</a> 
      * **PR #6916** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6916" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/6929?utm_source=stack-comment).

**Comment by connorgorman** [comment]
What about all the preexisting entries in the DB using the old format?

**Comment by janisz** [comment]
> What about all the pre existing entries in the DB using the old format?

I think they are fine. Queries will be the same a before and will work, when upsert will happen we will replace them with `{}`. I thought about using v4 approach and store `'null`' but this doesn't make much sense.

**Comment by connorgorman** [comment]
If we change all the values to {}, will pgx v4 still work the same way? We need to ensure any changes are completely backwards compatible

**Comment by janisz** [comment]
That's tricky. If we want be 100% backward compatible then we should store `'null'`.

The only incompatibility I could think of is a query that checks exact value or if value is a string and not map. I think our code does not allow such queries but I'm not an expert in this area. @md2119 do we expose any other queries than tested for maps?

**Comment by dashrews78** [comment]
> That's tricky. If we want be 100% backward compatible then we should store `'null'`.
> 
> The only incompatibility I could think of is a query that checks exact value or if value is a string and not map. I think our code does not allow such queries but I'm not an expert in this area. @md2119 do we expose any other queries than tested for maps?

We pretty much have to be backwards compatible in the event of a rollback from 4.2 to 4.1.  Greatly increases the amount of testing required to verify that we can make this change.

**Comment by janisz** [comment]
I added test for this behaviour 
- https://github.com/stackrox/stackrox/pull/6937

**Review by connorgorman — APPROVED** [review]

File: tools/generate-helpers/pg-table-bindings/store.go.tpl [hunk 1]
```diff
@@ -205,6 +205,8 @@ func {{ template "insertFunctionName" $schema }}({{ if eq (len $schema.Children)
         pgutils.NilOrTime({{$field.Getter "obj"}}),
         {{- else if eq $field.SQLType "uuid" }}
         pgutils.NilOrUUID({{$field.Getter "obj"}}),
+        {{- else if eq $field.DataType "map" }}
+        pgutils.EmptyOrMap({{$field.Getter "obj"}}),
```
Comment (rhybrillou) [inline tools/generate-helpers/pg-table-bindings/store.go.tpl:209]: I think the copyFrom template needs this change too (at least to be future-proof).
Maybe the `{{- range $field := $schema.DBColumnFields -}} ... {{- end }}` pattern could be extracted into another template.

**Review by rhybrillou — CHANGES_REQUESTED** [review]
please ensure value array construction consistency across `insertInto` and `copyFrom`

File: tools/generate-helpers/pg-table-bindings/store.go.tpl (hunk 1 above)
Comment (janisz) [reply tools/generate-helpers/pg-table-bindings/store.go.tpl:209]: oh, good point. I'll try to add a test for it too

Comment (janisz) [reply tools/generate-helpers/pg-table-bindings/store.go.tpl:209]: - https://github.com/stackrox/stackrox/pull/7108

**Review by rhybrillou — APPROVED** [review]

**Comment by janisz** [comment]
@janisz merged this pull request with [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/6929).

**Pull request merged**